### ISSUE TITLE
tensorflow op FusedBatchNorm is not supported

### ISSUE THREAD

**Issue by laobadao** [onnx/tensorflow-onnx#52]
I am trying to convert a yolov2 model，use my own frozen pb,In onnx‘s models , I saw a tiny-yolov2 model ,has a BatchNormalization Op, So  if  I convert this tf FusedBatchNorm Op to onnx BatchNormalization op ,  What should I need to pay special attention to?

**Comment by guschmue** [comment]
The converter should be folding batchnorms here
https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/tfonnx.py#L1187
but coincidentally I ran into the same issue with a yolo3 model.
Looking ...

**Comment by pengwa** [comment]
@guschmue ,  maybe the [fold_batch_norms (triggered by TransformGraph)](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/fold_batch_norms.cc) used in https://github.com/onnx/tensorflow-onnx/blob/8432455b5e2106ae1ca7bb7adcced7cd87f72afb/tf2onnx/tfonnx.py#L1187 is doing different thing with [fold_batch_norms](https://github.com/tensorflow/tensorflow/blob/b7300de4ef75bdd9373bccc4ae5b98135a70287b/tensorflow/python/tools/optimize_for_inference_lib.py#L204) ?

**Comment by pengwa** [comment]
please ignore my comments, just understood fold_old_batch_norms is replacing FusedBatchNorm.

**Comment by guschmue** [comment]
I think it gets triggered because I had disabled constant folding:
https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/tfonnx.py#L1186
Turning it on might take care of it but in my case I run into some known tf issue. Need to look some more.

**Comment by pengwa** [comment]
@guschmue enabling constant folding helps with my ResNetV1 152 model conversion. :) .

**Comment by laobadao** [comment]
@guschmue thanks a lot , it works,but some something wrong when I use run_caffe2 by running ResNet_V1_50 

run_onnx FAIL [enforce fail at operator.cc:113] schema->Verify(operator_def). Operator def did not pass schema checking: input: "resnet_v1_50/conv1/Conv2D__3:0" input: "resnet_v1_50/conv1/BatchNorm/gamma:0" input: "resnet_v1_50/conv1/BatchNorm/beta:0" input: "resnet_v1_50/conv1/BatchNorm/moving_mean:0" input: "resnet_v1_50/conv1/BatchNorm/moving_variance:0" output: "resnet_v1_50/conv1/BatchNorm/FusedBatchNorm:0" output: "resnet_v1_50/conv1/BatchNorm/FusedBatchNorm:1" output: "resnet_v1_50/conv1/BatchNorm/FusedBatchNorm:2" output: "resnet_v1_50/conv1/BatchNorm/FusedBatchNorm:3" output: "resnet_v1_50/conv1/BatchNorm/FusedBatchNorm:4" name: "resnet_v1_50/conv1/BatchNorm/FusedBatchNorm" type: "SpatialBN" arg { name: "epsilon" f: 1.001e-05 } arg { name: "is_test" i: 1 } device_option { device_type: 0 cuda_gpu_id: 0 } 
=== RESULT: 1 failed of 1, backend=caffe2

**Comment by guschmue** [comment]
It might be complaining about missing input. But for this to show there would need to be a onnx BatchNormalization operator in the graph which I do not expect. Also this must be after caffe2 translated the onnx graph to a caffe2 simplenet but I'm somewhat sure they run the onnx checker before that and onnx checker should have complained about a missing input.
Let me add some unit tests.
You can add the option ```--onnx-file /tmp/``` to the test app and than inspect the onnx graph with ```python tools/dumponnx.py --input /tmp/model.onnx``` ... might help to understand how we got here.

**Comment by laobadao** [comment]
@guschmue Thanks again, I used the dump-onnx.py tool to view the BatchNormalization Op, I found that because there is no conversion to the is_traing parameter in tensorflow FusedBN Op, converting to the is_test parameter of one of ONNX BatchNormalization， results in the default training mode, there are an additional 4 outputs, and I noticed that there are three BatchNormalizations in onnx. Different versions. There is an important is_test parameter to distinguish between training and testing modes. SpatialBN Op in caffe2 requires the is_test parameter, I guess I may modify my code based on this problem. what do you think?

**Comment by laobadao** [comment]
Finally, in resnet_v1_50, FusedBN is the training mode with 4 extra outputs. Now my approach is to delete the output, leaving only one Y, caffe2 no longer prompts the error, but a new error occurs.run_caffe2()

[enforce fail at conv_pool_op_base.h:154] pads_[dim] >= 0. -1 vs 0

my ckpt:

https://github.com/tensorflow/models/tree/master/research/slim#contacts ，

ResNet V1 50 --- resnet_v1_50_2016_08_28.tar.gz

frozen pb link:

链接: https://pan.baidu.com/s/1okICzfNKwbtjc6igGSnRoQ 密码: r47i

resnet_v1_50:
  model: tests/models/resnet_v1_50/pb_model/frozen_resnet_v1_50.pb
  input_get: get_beach
  inputs:
    "input:0": [1, 224, 224, 3]
  outputs:
    - resnet_v1_50/logits/BiasAdd:0

**Comment by guschmue** [comment]
I have added some unit test to make testing a little easier:
https://github.com/onnx/tensorflow-onnx/pull/63
I also added support for NHWC, possible that your pads issue goes away with this.
This works ok with our internal runtime.
caffe2 has issue ... need to add support for opset 1 and opset 6 as well.

**Comment by laobadao** [comment]
@guschmue thanks, FusedBN is totally OK  for me ,now  I have a another problem about tensorflow 
layers.max_pool2d(inputs, [1, 1], stride=[2,2], scope=scope)  to implement subsample in resnetv1_50, Some add_padding  problem appeared ,like  [-1,-1,0,0] I am  adding a new issue.
